daemon/mgr/container_list.go

```diff
@@ -0,0 +1,197 @@
+		return nil, err
+	}
+	return &filterContext{
+		condition:  option.Filter,
+		all:        option.All,
+		filterFunc: option.FilterFunc,
+	}, nil
+}
+
+// matchFilter filters value matchs field of condition
+func (fc *filterContext) matchFilter(field, value string) bool {
+	if value == "" {
+		return false
+	}
+
+	filters, exist := fc.condition[field]
+	if !exist || len(filters) == 0 {
+		return false
+	}
+
+	var (
+		match = false
+		err   error
+	)
+	for _, f := range filters {
+		if match, err = regexp.MatchString(f, value); err != nil {
+			continue
+		}
+		if match {
+			break
+		}
+	}
+
+	return match
+}
+
+// matchKVFilter filters map value matchs field of condition,
+// also support filter not equal
+func (fc *filterContext) matchKVFilter(field string, value map[string]string) bool {
+	filters, exist := fc.condition[field]
+	if !exist || len(filters) == 0 {
+		return false
+	}
+
+	// repeat value not support, eg: {"label": []string{"a=b", "a=c"}}, c will overwrite b
+	equalKV := make(map[string]string)
+	unequalKV := make(map[string]string)
+	for _, f := range filters {
+		if f == "" {
+			continue
+		}
+		splits := strings.SplitN(f, "!=", 2)
+		if len(splits) == 2 {
+			unequalKV[splits[0]] = splits[1]
+			continue
+		}
+
+		splits = strings.SplitN(f, "=", 2)
+		if len(splits) == 2 {
+			equalKV[splits[0]] = splits[1]
+			continue
+		}
+
+		equalKV[f] = ""
+	}
+
+	if len(equalKV) == 0 && len(unequalKV) == 0 {
+		return false
+	}
+
+	match := false
+	for k, v := range value {
+		// filter equal condition
+		if equalValue, exist := equalKV[k]; exist {
+			if equalValue == "" {
+				match = true
+				break
+			}
+			if equalValue == v {
+				match = true
+				break
+			}
+		}
+		// filter unequal condition
+		if unequalValue, exist := unequalKV[k]; exist && unequalValue != v {
+			match = true
+			break
+		}
+	}
+
+	return match
+}
+
+// filter does all select container work
+func (fc *filterContext) filter(c *Container) bool {
+	isNonStop := c.IsRunningOrPaused()
+
+	// if filterFunc is defined, we not try other filter condition
+	if fc.filterFunc != nil {
+		if match := fc.filterFunc(c); match {
+			// consider if need to add non-running container
+			return isNonStop || fc.all
+		}
+		return false
+	}
+
+	if len(fc.condition) == 0 {
+		// if no filter condition, skip filter process,
+		// add container according to flag `all` or running status
+		return isNonStop || fc.all
+	}
+
+	match := false
+	for name := range fc.condition {
+		switch name {
+		case labelFilter:
+			match = fc.matchKVFilter(labelFilter, c.Config.Labels)
+		case idFilter:
+			match = fc.matchFilter(idFilter, c.ID)
+		case nameFilter:
+			match = fc.matchFilter(nameFilter, c.Name)
+		case statusFilter:
+			match = fc.matchFilter(statusFilter, string(c.State.Status))
+		}
+
+		if match {
+			if name != statusFilter {
+				// consider if need to add non-running container
+				match = isNonStop || fc.all
+			}
+			break
```

Comment:
one match condition will stop this filter routine. what if people need both `-f label=site=daily` and `-f label=network!=overlay`

Reply:
Yes, one match will break loop, the function used to filter the match thing, so one condition match means it is chosed, 
it is what we want, we just want output matched thing.

Reply:
Modify the logic, jump until condition not equal, and add flag `jump` to determine if a condition is useful. Test case has add for this case.